### PR TITLE
IGNITE-15462 Replace usages of map of int to IntMap where possible in Calcite module.

### PR DESCRIPTION
**Pull request by vladErmakov07**

### CODE REVIEW

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/AbstractIgniteJoin.java [hunk 1]
```diff
@@ -304,9 +303,10 @@ protected boolean projectsLeft(RelCollation collation) {
         ImmutableIntList sourceKeys = left2Right ? joinInfo.leftKeys : joinInfo.rightKeys;
         ImmutableIntList targetKeys = left2Right ? joinInfo.rightKeys : joinInfo.leftKeys;
 
-        Map<Integer, Integer> keyMap = new HashMap<>();
+        Int2IntOpenHashMap keyMap = new Int2IntOpenHashMap(joinInfo.leftKeys.size());
+
```
Comment (AMashenkov) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/AbstractIgniteJoin.java:307]: This change make no sense, a Mappings.target() will box, then unbox int values, in anyway.

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/QueryTemplate.java [hunk 2]
```diff
@@ -104,7 +103,7 @@ public ExecutionPlan map(PlanningContext ctx) {
     private List<Fragment> replace(List<Fragment> fragments, Fragment fragment, List<Fragment> replacement) {
         assert !nullOrEmpty(replacement);
 
-        Map<Long, Long> newTargets = new HashMap<>();
+        Long2LongOpenHashMap newTargets = new Long2LongOpenHashMap();
```
Comment (AMashenkov) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/QueryTemplate.java:106]: `newTargets.put(remote.exchangeId(), fragment0.fragmentId());`
and
` Long newTargetId = newTargets.get(sender.exchangeId());`
use boxed values.

Comment (AMashenkov) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/QueryTemplate.java:106]: `Long newTargetId = newTargets.get(sender.exchangeId());` return boxed value.

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteProject.java [hunk 3]
```diff
@@ -135,7 +134,7 @@ public IgniteProject(RelInput input) {
         if (fieldCollations.isEmpty())
             return Pair.of(nodeTraits, List.of(in.replace(RelCollations.EMPTY)));
 
-        Map<Integer, Integer> targets = new HashMap<>();
+        Int2IntOpenHashMap targets = new Int2IntOpenHashMap();
```
Comment (AMashenkov) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteProject.java:137]: `Integer newIndex = targets.get(inFieldCollation.getFieldIndex());` below boxes value.
Seems, value can be null.

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/QueryTemplate.java (hunk 2 above)
Comment (ygerzhedovich) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/QueryTemplate.java:106]: we must revise all usages of new intMap, for example, few lines below we get Long and check in on null - it's wrong way.

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/util/RexUtils.java [hunk 4]
```diff
@@ -173,7 +172,7 @@ public static IndexConditions buildSortedIndexConditions(
 
         condition = RexUtil.toCnf(builder(cluster), condition);
 
-        Map<Integer, List<RexCall>> fieldsToPredicates = mapPredicatesToFields(condition, cluster);
+        Int2ObjectOpenHashMap<List<RexCall>> fieldsToPredicates = mapPredicatesToFields(condition, cluster);
```
Comment (AMashenkov) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/util/RexUtils.java:175]: `ImmutableIntList.copyOf(fieldsToPredicates.keySet())` below lead to box/unbox values.
Let's replace with `ImmutableIntList.of(fieldsToPredicates.keySet().toIntArray)`

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/QueryTemplate.java (hunk 2 above)
Comment (vladErmakov07) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/QueryTemplate.java:106]: Revised all changes of Int and Long maps. In places with null check I replaced it with "containsKey" function.

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/AbstractIgniteJoin.java (hunk 1 above)
Comment (vladErmakov07) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/AbstractIgniteJoin.java:307]: Agree, reverted it.

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteMergeJoin.java [hunk 5]
```diff
@@ -166,30 +168,39 @@ else if (isPrefix(rightCollation.getKeys(), joinInfo.rightKeys))// preserve righ
...
         List<Integer> newLeftCollation, newRightCollation;
 
-        Map<Integer, Integer> leftToRight = joinInfo.pairs().stream()
-            .collect(Collectors.toMap(p -> p.source, p -> p.target));
+        Int2IntOpenHashMap leftToRight = new Int2IntOpenHashMap(pairs.size());
+
+        for (IntPair pair : pairs)
+            leftToRight.put(pair.source, pair.target);
 
         if (isPrefix(collationLeftPrj, joinInfo.leftKeys)) { // preserve collation
```
Comment (AMashenkov) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteMergeJoin.java:196]: Commons.isPrefix created HashSet with boxed values underneath.

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteMergeJoin.java [hunk 6]
```diff
@@ -166,30 +168,39 @@ else if (isPrefix(rightCollation.getKeys(), joinInfo.rightKeys))// preserve righ
...
             newLeftCollation = new ArrayList<>();
             newRightCollation = new ArrayList<>();
```
Comment (AMashenkov) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteMergeJoin.java:198]: ArrayList -> IgniteIntList ?

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteMergeJoin.java (hunk 5 above)
Comment (ygerzhedovich) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteMergeJoin.java:196]: I am sure that should be done as a separate task - replace for Set and List

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteMergeJoin.java (hunk 6 above)
Comment (ygerzhedovich) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteMergeJoin.java:198]: Let's skip it for current PR

**Review by AMashenkov — APPROVED** [review]

**Pull request merged**